README.md
[![ko-fi](https://www.ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/U7U1RSV5)

<a href="https://top.gg/bot/682946560417333283" >
  <img src="https://top.gg/api/widget/status/682946560417333283.svg" alt="Coronavirus COVID-19" />
</a>

[![PyPI pyversions](https://img.shields.io/badge/python-3.9-blue)](https://www.python.org/)  ![AppVeyor](https://img.shields.io/appveyor/ci/takitsu21/WarframeTrader) [![Discord](https://img.shields.io/discord/556268083681951759?color=blue&label=discord)](http://discord.gg/wTxbQYb) ![GitHub pull requests](https://img.shields.io/github/issues-pr/takitsu21/covid-19-tracker) ![GitHub issues](https://img.shields.io/github/issues/takitsu21/covid-19-tracker) [![GitHub](https://img.shields.io/github/license/takitsu21/covid-19-tracker)](LICENCE)
<p align="left"><a href="https://discordbots.org/bot/682946560417333283" >
  <img src="https://top.gg/api/widget/682946560417333283.svg?usernamecolor=FFFFFF&topcolor=000000" alt="Coronavirus COVID-19" />
</a></p>

# Coronavirus COVID-19

Get the latest update with notification alerts and graphics stats about COVID-19! For every country and region! World news are available too!

## Commands

Prefix : **`c! or @mention`**

**`<something>`** is required
**`[something]`** is optional
**`arg1 / arg2`** mean arg1 or arg2

| Command | Description |
|-------|-----------|
| **`c!help`** | Displays help command. |
| **`c!info`** | Displays every confirmed case. |
| **`c!list`** | List every country affected by COVID-19 |
| **`c!<s / stats> [log / country / log [country]]`** | Displays graphical statistics. If no args provided return linear graph for total cases. You can find countries with **full name** or **[ISO-3166-1](https://fr.wikipedia.org/wiki/ISO_3166-1)**.__Examples__ : `c!stats us`, `c!s log usa`, `c!stats log` |
| **`c!<r / region> <state/province / all> in <country>`** | Supported countries (**China, Canada, United States, Australia, Cruise Ship**). Displays regions infected within a  country or in the whole state with the `all` arg. The `in` (mandatory symbol) is interpreted as separator between the country and the region/province so don't forget it. Example 1 : `c!r new york in us`. Example 2 : `c!region all in china`. |
| **`c!track <country / [disable]>`** | Track country (bot will DM you update). __Examples__ : `c!track us`, `c!track disable`. |
| **`c!news`** | Views recent news about COVID-19 (update every 1 hour). |
| **`c!notification <country / disable> <every NUMBER> <hours / days / weeks>`** | (Only administrator) When new data is found, the bot will send you a notification where you executed the command, **server only**. __Examples__ : `c!notification usa every 3 hours`, `c!notification disable` |
| **`c!source`** | Displays source data which the bot is based on. |
| **`c!suggestion <message>`** | Send suggestion feedback. |
| **`c!bug <message>`** | Send bug feedback. |
| **`c!about`** | Displays information about the bot. |
| **`c!invite`** | Displays bot link invite. |
| **`c!vote`** | Displays bot vote link. |

## Few examples

- `c!s us`

![Example Stats](https://i.imgur.com/mykjbCr.png)

- `c!s log`

![Stats log](https://i.imgur.com/XKKXRE2.png)

- `c!notif usa every 2 hours`

![Example notification](https://i.imgur.com/TJOUp00.png)

- `c!region all in us`

![State/Province](https://i.imgur.com/4L4R0FZ.png)

- `c!info`

![Example info](https://i.imgur.com/LIIoU2O.png)

## Built With

* [discord.py]

## Author

* [**takitsu21**]

### License

This project is licensed under the MIT License - see the [LICENSE.md] file for details.

### Deployment

* Scaleway

# Changelog

## 0.0.8 - 2021-04-15

### Changed

- Improve charts for `c!daily` command which is prettier and more accurate [(#19)](https://github.com/takitsu21/covid-19-tracker/pull/19).
- Cooldown command prettier [(#17)](https://github.com/takitsu21/covid-19-tracker/pull/17).

### Fixed

- Fixed minor typos & grammatical errors [(#18)](https://github.com/takitsu21/covid-19-tracker/pull/18)

## 0.0.7 - 2021-04-07

### Fixed

- Fixed minor typos
- Fixed minor grammatical errors

## 0.0.6 - 2020-04-07

### Added

- `c!list` -> list every country that can be called by bot commands.
- **[ISO-3166-1](https://fr.wikipedia.org/wiki/ISO_3166-1)** for all countries

### Fixed

- Improve bot stability.
- Fixed minor issues.
- Fixed typos.

### Changed

- `c!notification <enable | disable>` -> `c!nofitication <country | disable> <every NUMBER> <hours | days | weeks>` you can have a notification about a country in a specific channel of your server.
- `c!stats [log]` -> `c!<s | stats> [log | country | log [country]]` you can now have a log graph for countries.

- Graph display Total confirmed - active.

## 0.0.5 - 2020-03-29

### Added

- Now with `c!stats <country>` you got a minimap in the right top corner about the selected country.

### Fixed

- Prefix can be interpreted as `C!` and `c!` (useful when you are on mobile).

### Changed

- Back to old format according to your suggestion.

## 0.0.4 - 2020-03-27

### Added

- New emote.
- More informations in `c!source`.
- Add Total active cases in `c!stats`

### Fixed

- Fix Notification and tracker.
- Improve data stability.

### Changed

- `c!stats` changed to -> `c!stats [country | log]` and can now show specific graphic for each countries, you can also
  get the logarithmic graphic.
- change `c!help` for `c!stats`.

## 0.0.3 - 2020-03-17

### Added

- New command : `c!<r | region> <STATE/PROVINCE | all> in <COUNTRY>`, now you can get stats from state/province in
  supported countries
  (**China, Canada, United States, Australia, Cruise Ship**).
- New field in `c!help` for `c!<r | region>` command.
- [Patreon] field added in `c!about`.
- World Health Organization (WHO) source in `c!source`.
- Different formatting on mobile and desktop
- New bot icon

### Fixed

- English typos.

### Changed

- Better formatting visualization for `c!country`, `c!info`, `c!notication`, `c!track`
- Better graphic visualization on x and y axis, timeline rotation and some color changes.
- Embed color

## 0.0.2 - 2020-03-13

### Added

- `c!news` recent news about COVID-19 using newsapi.org
- New field in `c!help` for `c!news`

### Fixed

- Fix typos that didn't match with the current difference visualisation.
- Fix GMT update typos.
- Improvements on data update.

## 0.0.1 - 2020-03-12

### Added

- `c!vote`
- `c!track <COUNTRY | [disabled]>`
- `c!suggestion <MESSAGE>`
- `c!bug <MESSAGE>`
- New data sources.

### Fixed

- Graphical statistics now update the current day cases.
- You can use country code to get a country data with `c!country <COUNTRY>`

[Patreon]: https://www.patreon.com/takitsu

[Changelog]: CHANGELOG

[discord.py]: https://discordpy.readthedocs.io/en/latest/

[**takitsu21**]: https://github.com/takitsu21/

[LICENSE.md]: LICENCE



CHANGELOG.md
# Changelog

## 0.0.9 - 2021-07-01

### Added

- Slash commands availaible !

### Changed

- Python 3.8 to Python 3.9

### Fixed

- Minor fixes

## 0.0.8 - 2021-04-15

### Changed

- Improve charts for `c!daily` command which is prettier and more accurate.
- Cooldown command prettier [(#17)](https://github.com/takitsu21/covid-19-tracker/pull/17).

### Fixed

- Fixed minor typos & grammatical errors [(#18)](https://github.com/takitsu21/covid-19-tracker/pull/18)

## 0.0.7 - 2020-04-11

### Added

- `c!setprefix <new_prefix>` You can now set your own guild prefix.
- `@Coronavirus COVID-19 getprefix` Views current guild prefix.

### Fixed

- Fix spelling errors, thanks to [benhovinga](https://github.com/benhovinga) for contributing.

### Changed

- cured to recovered in `c!country` [Issue #8](https://github.com/takitsu21/covid-19-tracker/issues/8)

## 0.0.6 - 2020-04-07

### Added

- `clist` -> list every country that can be call by bot commands.
- **[ISO-3166-1](https://fr.wikipedia.org/wiki/ISO_3166-1)** for all countries

### Fixed

- Improve bot stability.
- Fix minor issues.
- Fix typos.

### Changed

- `c!notification <enable | disable>` -> `c!nofitication <country | disable> <every NUMBER> <hours | days | weeks>` you can have notification about country in a specific channel for your server.
- `c!stats [log]` -> `c!<s | stats> [log | country | log [country]]` you can now have log graph for countries.
- Graph display Total confirmed - active.

## 0.0.5 - 2020-03-29

### Added

- Now with `c!stats <country>` you got a minimap in the right top corner about the selected country.

### Fixed

- Prefix can be interpreted as `C!` and `c!` (useful when you are on mobile).

### Changed

- Back to old format according to your suggestion.

## 0.0.4 - 2020-03-27

### Added

- New emote.
- More information in `c!source`.
- Add Total active cases in `c!stats`

### Fixed

- Fix Notification and tracker.
- Improve data stability.

### Changed

- `c!stats` changed to -> `c!stats [country | log]` and can now show specific graphic for each countries, you can also get the logarithmic graphic.
- change `c!help` for `c!stats`.


## 0.0.3 - 2020-03-17

### Added

- New command : `c!<r | region> <STATE/PROVINCE | all> in <COUNTRY>`,
  now you can get stats from state/province in supported countries
  (**China, Canada, United States, Australia, Cruise Ship**).
- New field in `c!help` for `c!<r | region>` command.
- [Patreon] field added in `c!about`.
- World Health Organization (WHO) source in `c!source`.
- Different formatting on mobile and desktop
- New bot icon

### Fixed

- English typos.

### Changed

- Better formatting visualization for `c!country`, `c!info`, `c!notication`, `c!track`
- Better graphic visualization on x and y axis, timeline rotation and some color changes.
- Embed color

## 0.0.2 - 2020-03-13

### Added

- `c!news` recent news about COVID-19 using newsapi.org
- New field in `c!help` for `c!news`

### Fixed

- Fix typos that didn't match with the current difference visualisation.
- Fix GMT update typos.
- Improvements on data update.

## 0.0.1 - 2020-03-12

### Added

- `c!vote`
- `c!track <COUNTRY | [disabled]>`
- `c!suggestion <MESSAGE>`
- `c!bug <MESSAGE>`
- New data sources.

### Fixed

- Graphical statistics now update the current day cases.
- You can use country code to get a country data with `c!country <COUNTRY>`

[Patreon]: https://www.patreon.com/takitsu

main.py
import datetime
import logging
import os

import aiomysql
import discord
from aiohttp import ClientSession
from decouple import config
from discord.ext import commands
from discord.ext.commands import when_mentioned_or
from discord.utils import find
from discord_slash import SlashCommand

import src.utils as utils
from src.database import Pool

logger = logging.getLogger('covid-19')
logger.setLevel(logging.DEBUG)
handler = logging.FileHandler(
    filename='covid-19.log',
    encoding='utf-8',
    mode='w'
)
handler.setFormatter(logging.Formatter(
    '%(asctime)s:%(levelname)s:%(name)s: %(message)s'
)
)
logger.addHandler(handler)


class Covid(commands.AutoShardedBot, Pool):

    def __init__(self, *args, **kwargs):
        super().__init__(
            command_prefix=self._get_prefix,
            activity=discord.Game(name="c!help | Loading shards..."),
            status=discord.Status.dnd
        )
        super(Pool, self).__init__()
        self.slash = SlashCommand(
            self,
            sync_commands=True,
            sync_on_cog_reload=True
        )
        self.debug_token = kwargs.get("debug_token", False)
        self.remove_command("help")
        self._load_extensions()
        self.news = None
        self.http_session = None
        self.pool = None
        self.auto_update_running = False
        self.script_start_dt = datetime.datetime.now()
        self.thumb = "https://upload.wikimedia.org/wikipedia/commons/thumb/2/26/COVID-19_Outbreak_World_Map.svg/langfr-1000px-COVID-19_Outbreak_World_Map.svg.png?t="
        self.author_thumb = "https://upload.wikimedia.org/wikipedia/commons/thumb/e/ef/International_Flag_of_Planet_Earth.svg/1200px-International_Flag_of_Planet_Earth.svg.png"
        self.loop.create_task(self.startup_tasks())

    async def _get_prefix(self, bot, message):
        try:
            prefix = await self.getg_prefix(message.guild.id)
        except:
            if message.content[0:2] == "C!":
                prefix = "C!"
            else:
                prefix = "c!"
        return when_mentioned_or(prefix)(bot, message)

    def _load_extensions(self):
        for file in os.listdir("cogs/"):
            try:
                if file.endswith(".py"):
                    self.load_extension(f'cogs.{file[:-3]}')
                    logger.info(f"{file} loaded")
            except Exception:
                logger.exception(f"Fail to load {file}")

    def _unload_extensions(self):
        for file in os.listdir("cogs/"):
            try:
                if file.endswith(".py"):
                    self.unload_extension(f'cogs.{file[:-3]}')
                    logger.info(f"{file} unloaded")
            except Exception:
                logger.exception(f"Fail to unload {file}")

    async def on_command_error(self, ctx, error):
        if isinstance(error, commands.CommandNotFound):
            return await ctx.send(
                "That command does not exist :confused:\nPlease use `{}help` for a list of commands".format(
                    ctx.prefix)
            )
            # Handling Command Not Found Errors
        if isinstance(error, commands.CommandOnCooldown):
            await ctx.send('{}'.format(ctx.author.mention), embed=discord.Embed(
                title=":alarm_clock: Cooldown Error",
                description='Please cooldown a little, try again in `{:.2}s`'.format(
                    error.retry_after),
                color=utils.COLOR), delete_after=5)
            # A little artistic touch won't hurt, I'll attach a screenshot in the PR description
        else:
            # raise error
            embed = discord.Embed(
                title="Error",
                description=f"{error} Invalid command see `c!help`.\nIf you think that is a bot error, report this issue on my discord [here](https://discordapp.com/invite/wTxbQYb) thank you.",
                timestamp=datetime.datetime.utcnow(),
                color=utils.COLOR
            )
            await ctx.send(embed=embed)

    async def on_guild_join(self, guild: discord.Guild):
        await self.wait_until_ready()
        chan_logger = self.get_channel(692815717078270052)
        try:
            general = find(lambda x: x.name == "general", guild.text_channels)
            if general and general.permissions_for(guild.me).send_messages:
                embed = discord.Embed(
                    description="You can support me on <:kofi:693473314433138718>[Kofi](https://ko-fi.com/takitsu) and vote on [top.gg](https://top.gg/bot/682946560417333283/vote) for the bot. <:github:693519776022003742> [Source code](https://github.com/takitsu21/covid-19-tracker)",
                    timestamp=utils.discord_timestamp(),
                    color=utils.COLOR
                )
                embed.set_author(
                    name="Coronavirus COVID-19 Tracker", icon_url=guild.me.avatar_url)
                embed.set_thumbnail(url=self.thumb)
                embed.add_field(name="Vote",
                                value="[Click here](https://top.gg/bot/682946560417333283/vote)")
                embed.add_field(name="Invite Coronavirus COVID-19",
                                value="[Click here](https://discordapp.com/oauth2/authorize?client_id=682946560417333283&scope=bot&permissions=313408)")
                embed.add_field(name="Discord Support",
                                value="[Click here](https://discordapp.com/invite/wTxbQYb)")
                embed.add_field(
                    name="Source code", value="[Click here](https://github.com/takitsu21/covid-19-tracker)")
                embed.add_field(name="Help command", value="c!help")
                embed.add_field(name="Prefix", value="c!")
                nb_users = 0
                channels = 0
                for s in self.guilds:
                    nb_users += len(s.members)
                    channels += len(s.channels)
                embed.add_field(name="<:confirmed:688686089548202004> Confirmed",
                                value=self._data["total"]["confirmed"])
                embed.add_field(name="<:recov:688686059567185940> Recovered",
                                value=self._data["total"]["recovered"])
                embed.add_field(
                    name="<:_death:688686194917244928> Deaths", value=self._data["total"]["deaths"])
                embed.add_field(
                    name="<:servers:693053697453850655> Servers", value=len(self.guilds))
                embed.add_field(
                    name="<:users:693053423494365214> Members", value=nb_users)
                embed.add_field(
                    name="<:hashtag:693056105076621342> Channels", value=channels)
                embed.add_field(name="<:stack:693054261512110091> Shards",
                                value=f"{guild.shard_id + 1}/{self.bot.shard_count}")
                embed.set_footer(text="Made by Taki#0853 (WIP) " + utils.last_update(utils.DATA_PATH),
                                 icon_url=guild.me.avatar_url)
                await general.send(embed=embed)
        except:
            pass

        embed = discord.Embed(
            title="Bot added to " + guild.name,
            timestamp=datetime.datetime.utcnow(),
            color=utils.COLOR
        )
        embed.add_field(
            name="<:users:693053423494365214> Members",
            value=len(guild.members)
        )
        embed.add_field(
            name="<:hashtag:693056105076621342> Channels",
            value=len(guild.channels)
        )
        embed.set_thumbnail(url=guild.icon_url)
        embed.set_footer(icon_url=guild.me.avatar_url)
        await chan_logger.send(embed=embed)

    async def on_guild_remove(self, guild: discord.Guild):
        try:
            await self.delete_notif(guild.id)
        except:
            pass
        try:
            await self.delete_prefix(guild.id)
        except:
            pass

    async def startup_tasks(self):
        if self.http_session is None:
            self.http_session = ClientSession(loop=self.loop)
        if self.pool is None:
            try:
                self.pool = await aiomysql.create_pool(
                    host=config("db_host"),
                    port=3306,
                    user=config("db_user"),
                    password=config("db_token"),
                    db=config("db_user"),
                    maxsize=10,
                    loop=self.loop,
                    autocommit=True
                )
                logger.info("pool created")
            except Exception as e:
                logger.exception(e, exc_info=True)

    async def on_ready(self):
        await self.startup_tasks()
        await self.change_presence(
            activity=discord.Game(
                name=f"c!help | coronavirus.jessicoh.com/api/"
            )
        )

    async def on_message(self, message):
        if message.author.bot:
            return
        return await self.process_commands(message)

    def run(self, *args, **kwargs):
        try:
            super().run(config("debug") if self.debug_token else config("token"), *args, **kwargs)
        except KeyboardInterrupt:
            try:
                self.loop.run_until_complete(self._close())
                self.loop.run_until_complete(self.http_session.close())
                logger.info("Shutting down")
                exit(0)
            except Exception as e:
                logger.exception(e, exc_info=True)
                exit(1)

    @property
    def topgg(self):
        return self.get_cog("TopGG").topgg

    @property
    def continent_code(self):
        return self.get_cog("Statistics").continent_code


if __name__ == "__main__":
    import sentry_sdk
    from sentry_sdk.integrations.aiohttp import AioHttpIntegration

    sentry_sdk.init(
        config("SENTRY_TOKEN"),
        traces_sample_rate=1.0,
        integrations=[AioHttpIntegration()]
    )
    bot = Covid()
    bot.run(reconnect=True)
